packages/gatsby-plugin-sharp/src/process-file.js

```diff
@@ -73,139 +74,130 @@ const argsWhitelist = [
-      })
-      .png({
-        compressionLevel: args.pngCompressionLevel,
-        adaptiveFiltering: false,
-        force: args.toFormat === `png`,
-      })
-      .webp({
-        quality: args.webpQuality || args.quality,
-        force: args.toFormat === `webp`,
-      })
-      .tiff({
-        quality: args.quality,
-        force: args.toFormat === `tiff`,
-      })
-
-    // jpeg
-    if (!options.useMozJpeg) {
-      clonedPipeline = clonedPipeline.jpeg({
-        quality: args.jpegQuality || args.quality,
-        progressive: args.jpegProgressive,
-        force: args.toFormat === `jpg`,
-      })
+    try {
+      const { outputPath, args } = transform
+      debug(`Start processing ${outputPath}`)
+      await fs.ensureDir(path.dirname(outputPath))
+
+      let clonedPipeline = transforms.length > 1 ? pipeline.clone() : pipeline
+
+      if (args.trim) {
+        clonedPipeline = clonedPipeline.trim(args.trim)
+      }
+
+      if (!args.rotate) {
+        clonedPipeline = clonedPipeline.rotate()
+      }
+
+      // Sharp only allows ints as height/width. Since both aren't always
+      // set, check first before trying to round them.
+      let roundedHeight = args.height
+      if (roundedHeight) {
+        roundedHeight = Math.round(roundedHeight)
+      }
+
+      let roundedWidth = args.width
+      if (roundedWidth) {
+        roundedWidth = Math.round(roundedWidth)
+      }
+
+      clonedPipeline
+        .resize(roundedWidth, roundedHeight, {
+          position: args.cropFocus,
+          fit: args.fit,
+          background: args.background,
+        })
+        .png({
+          compressionLevel: args.pngCompressionLevel,
+          adaptiveFiltering: false,
+          force: args.toFormat === `png`,
+        })
+        .webp({
+          quality: args.webpQuality || args.quality,
+          force: args.toFormat === `webp`,
+        })
+        .tiff({
+          quality: args.quality,
+          force: args.toFormat === `tiff`,
+        })
+
+      // jpeg
+      if (!options.useMozJpeg) {
+        clonedPipeline = clonedPipeline.jpeg({
+          quality: args.jpegQuality || args.quality,
+          progressive: args.jpegProgressive,
+          force: args.toFormat === `jpg`,
+        })
+      }
+
+      // grayscale
+      if (args.grayscale) {
+        clonedPipeline = clonedPipeline.grayscale()
+      }
+
+      // rotate
+      if (args.rotate && args.rotate !== 0) {
+        clonedPipeline = clonedPipeline.rotate(args.rotate)
+      }
+
+      // duotone
+      if (args.duotone) {
+        clonedPipeline = await duotone(
+          args.duotone,
+          args.toFormat,
+          clonedPipeline
+        )
+      }
+
+      // lets decide how we want to save this transform
+      if (args.toFormat === `png`) {
+        await compressPng(clonedPipeline, outputPath, {
+          ...args,
+          stripMetadata: options.stripMetadata,
+        })
+        return transform
```

Comment:
does returning `transform` here cause it to be retained in as job's "result"? If so - do we need those transforms as a job's result?

Reply:
Unsure but the memory issues are fixed with latest commit